<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Import-necessary-librairies" data-toc-modified-id="Import-necessary-librairies-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Import necessary librairies</a></span></li><li><span><a href="#Import-CSV-data-files-into-Python" data-toc-modified-id="Import-CSV-data-files-into-Python-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Import CSV data files into Python</a></span></li><li><span><a href="#Contatenate-all-csv-files-in-one-dataframe" data-toc-modified-id="Contatenate-all-csv-files-in-one-dataframe-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Contatenate all csv files in one dataframe</a></span></li><li><span><a href="#Analyze-the-structure-of-dataframe-and-perform-the-data-cleaning" data-toc-modified-id="Analyze-the-structure-of-dataframe-and-perform-the-data-cleaning-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Analyze the structure of dataframe and perform the data cleaning</a></span><ul class="toc-item"><li><span><a href="#Add-new-columns" data-toc-modified-id="Add-new-columns-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Add new columns</a></span></li><li><span><a href="#Drop-duplicated-rows" data-toc-modified-id="Drop-duplicated-rows-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Drop duplicated rows</a></span></li><li><span><a href="#Move-the-categories-in-the-column-'Specie'-to-the-level-1-of-indexes-of-columns" data-toc-modified-id="Move-the-categories-in-the-column-'Specie'-to-the-level-1-of-indexes-of-columns-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Move the categories in the column 'Specie' to the level 1 of indexes of columns</a></span></li><li><span><a href="#Convert-the-values-in-the-column-'Date'-into-the-type-'Date-time'" data-toc-modified-id="Convert-the-values-in-the-column-'Date'-into-the-type-'Date-time'-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Convert the values in the column 'Date' into the type 'Date time'</a></span></li><li><span><a href="#Analyse-and-drop-unnecessary-columns" data-toc-modified-id="Analyse-and-drop-unnecessary-columns-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Analyse and drop unnecessary columns</a></span></li><li><span><a href="#Analyse-missing-values" data-toc-modified-id="Analyse-missing-values-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Analyse missing values</a></span></li></ul></li><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Exploratory Data Analysis</a></span></li></ul></div>

### Import necessary librairies

In [469]:
pip install pycountry-convert #Install the library used to convert value between country code, country name and continent name.

Note: you may need to restart the kernel to use updated packages.


ERROR: Invalid requirement: '#Install'


In [470]:
pip install pytest-filter-subpackage

Note: you may need to restart the kernel to use updated packages.


In [673]:
import pandas as pd
import numpy as np
import pycountry # to get the country name from country code
import pycountry_convert as pc # to get the continent name from country code
import requests #to get data from the internet
from io import StringIO # library helps read and write a string as files

### Import CSV data files into Python

We store all csv files in google drive. We use a function to collect data from urls of google drive and store them in one dataframe.

In [750]:
# Function to get the data from google drive

def get_data(org_url):
    file_id = org_url.split('/')[-2]
    dwn_url='https://drive.google.com/uc?export=download&id=' + file_id
    url = requests.get(dwn_url).text
    csv_raw = StringIO(url)
    data = pd.read_csv(csv_raw) #latin1  
    return data

In [752]:
# List of urls link to the csv files in google drive

org_url1 = 'https://drive.google.com/file/d/1X8wWfyMJm1VzuCfXMMBzb-UmzJtG1JOC/view?usp=sharing'
org_url2 = 'https://drive.google.com/file/d/1uY04U1DoQM5jl4AI5NmXBI63GPvjp38j/view?usp=sharing'
org_url3 = 'https://drive.google.com/file/d/1V4kpWQ9cHthntyyHFOU_XCMq586TgOX5/view?usp=sharing'
org_url4 = 'https://drive.google.com/file/d/1H5P4a678ig-JWGe7gGZgyqNT9j1nwkKj/view?usp=sharing'
org_url5 = 'https://drive.google.com/file/d/1UuqZuCT9urSOfm0ZribIzFAK-baJVXAL/view?usp=sharing'

lst_url = [org_url1,org_url2,org_url3,org_url4,org_url5]

In [753]:
# Use the get_data function to collect the data and store them in a dataframe

df = []

for n in lst_url:
    data = get_data(n)
    df.append(data)
    
df = pd.concat(df)

df.head()

Date Country       City Specie  count    min    max  median  variance
0  1/16/2019      AE  Abu Dhabi   pm10     24   86.0   99.0    97.0    179.40
1  1/22/2019      AE  Abu Dhabi   pm10     24   51.0   57.0    55.0     23.75
2  1/26/2019      AE  Abu Dhabi   pm10     24  136.0  173.0   160.0    941.96
3   1/7/2019      AE  Abu Dhabi   pm10     24   60.0   91.0    72.0   1006.88
4  1/10/2019      AE  Abu Dhabi   pm10     24   82.0   93.0    87.0     57.97

In [754]:
df['City'].unique()

array(['Abu Dhabi', 'Dubai', 'Prague', 'Ostrava', 'Brno', 'Pilsen',
       'Olomouc', 'Singapore', 'Salzburg', 'Linz', 'Innsbruck', 'Vienna',
       'Graz', 'Liège', 'Brussels', 'Charleroi', 'Gent', 'Namur',
       'Antwerpen', 'Jeonju', 'Seoul', 'Busan', 'Sejong', 'Ulsan',
       'Yeosu', 'Gwangju', 'Seongnam-si', 'Daejeon', 'Pohang', 'Changwon',
       'Suwon', 'Cheongju-si', 'Chuncheon', 'Daegu', 'Jeju City',
       'Incheon', 'Suncheon', 'Mokpo', 'Andong', 'Ulan Bator',
       'Trondheim', 'Oslo', 'Stavanger', 'Bergen', 'Kathmandu', 'Pokhara',
       'Lima', 'Sarajevo', 'Zenica', 'Tuzla', 'São José dos Campos',
       'São Paulo', 'Tel Aviv', 'Haifa', 'Jerusalem', 'Ashdod',
       'Netanya', 'Ashkelon', 'Petaáº\x96 Tiqwa', 'New Delhi',
       'Hyderabad', 'Delhi', 'Chandigarh', 'Bhopal', 'Lucknow',
       'GhÄ\x81ziÄ\x81bÄ\x81d', 'HÄ\x81pur', 'Gandhinagar', 'Chennai',
       'Nashik', 'Visakhapatnam', 'Mumbai', 'Jaipur', 'Patna',
       'Muzaffarnagar', 'Thiruvananthapuram', 'B

In [549]:
# #Our dataset is store in two seperated csv files

# df2020 = pd.read_csv('D:/Documents/Ironhack/Projet 3/DataAQI/waqi-covid19-airqualitydata-2020.csv')
# df2019Q1 = pd.read_csv('D:/Documents/Ironhack/Projet 3/DataAQI/waqi-covid19-airqualitydata-2019Q1.csv')
# df2019Q2 = pd.read_csv('D:/Documents/Ironhack/Projet 3/DataAQI/waqi-covid19-airqualitydata-2019Q2.csv')
# df2019Q3 = pd.read_csv('D:/Documents/Ironhack/Projet 3/DataAQI/waqi-covid19-airqualitydata-2019Q3.csv')
# df2019Q4 = pd.read_csv('D:/Documents/Ironhack/Projet 3/DataAQI/waqi-covid19-airqualitydata-2019Q4.csv')

### Contatenate all csv files in one dataframe

In [550]:
# list_df = [df2020,df2019Q1,df2019Q2,df2019Q3,df2019Q4]
# df = pd.concat(list_df, ignore_index = True)

### Analyze the structure of dataframe and perform the data cleaning

In [675]:
df.head()

Date Country       City Specie  count    min    max  median  variance
0  1/16/2019      AE  Abu Dhabi   pm10     24   86.0   99.0    97.0    179.40
1  1/22/2019      AE  Abu Dhabi   pm10     24   51.0   57.0    55.0     23.75
2  1/26/2019      AE  Abu Dhabi   pm10     24  136.0  173.0   160.0    941.96
3   1/7/2019      AE  Abu Dhabi   pm10     24   60.0   91.0    72.0   1006.88
4  1/10/2019      AE  Abu Dhabi   pm10     24   82.0   93.0    87.0     57.97

In [676]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3223468 entries, 0 to 1048570
Data columns (total 9 columns):
 #   Column    Dtype  
---  ------    -----  
 0   Date      object 
 1   Country   object 
 2   City      object 
 3   Specie    object 
 4   count     int64  
 5   min       float64
 6   max       float64
 7   median    float64
 8   variance  float64
dtypes: float64(4), int64(1), object(4)
memory usage: 245.9+ MB


In [677]:
df.shape

(3223468, 9)

As we can see, the dataframe has 3 223 468 rows which are the same number as the total rows in these csv files.

#### Add new columns

Based on the country code, we will add two new columns: Country_name and Continent for the sake of our further analysis

In [734]:
# by using the function pycountry, we create a dictionary that contains pairs of code and name of each country

countries = {}
for country in pycountry.countries:
    countries[country.name] = country.alpha_2

# We define a function to get the name of country based on the country code
def get_country_name(x):
        for name, code in countries.items(): 
            if code == x:
                return name

# We define a function to get the name of continent based on the country code
def get_continent(x):
    continent_code = pc.country_alpha2_to_continent_code(x)
    continent_name = pc.convert_continent_code_to_continent_name(continent_code)
    return continent_name

# We apply these functions above to create the 2 new columns called "Country_name" and "Continent"
df['Country_name'] = df['Country'].apply(lambda row: get_country_name(row))
df['Continent'] = df['Country'].apply(lambda row: get_continent(row))
df = df[['Date','Continent','Country','Country_name','City','Specie','count','min','max','median','variance']]

KeyError: "['Specie'] not in index"

In [679]:
df.head()

Date Continent Country          Country_name       City Specie  count  \
0  1/16/2019      Asia      AE  United Arab Emirates  Abu Dhabi   pm10     24   
1  1/22/2019      Asia      AE  United Arab Emirates  Abu Dhabi   pm10     24   
2  1/26/2019      Asia      AE  United Arab Emirates  Abu Dhabi   pm10     24   
3   1/7/2019      Asia      AE  United Arab Emirates  Abu Dhabi   pm10     24   
4  1/10/2019      Asia      AE  United Arab Emirates  Abu Dhabi   pm10     24   

     min    max  median  variance  
0   86.0   99.0    97.0    179.40  
1   51.0   57.0    55.0     23.75  
2  136.0  173.0   160.0    941.96  
3   60.0   91.0    72.0   1006.88  
4   82.0   93.0    87.0     57.97

#### Drop duplicated rows

In [680]:
df[df.duplicated()]

Date Continent Country Country_name     City    Specie  count  \
4        10/1/2019    Europe      SK     Slovakia  Košice       dew     72   
12       10/2/2019    Europe      SK     Slovakia  Košice       dew     66   
40       10/5/2019    Europe      SK     Slovakia  Košice       dew     23   
49       10/4/2019    Europe      SK     Slovakia  Košice       dew     24   
62       9/30/2019    Europe      SK     Slovakia  Košice       dew     60   
...            ...       ...     ...          ...      ...       ...    ...   
1048252   1/5/2020    Europe      UA      Ukraine     Kyiv  pressure    192   
1048415   1/4/2020    Europe      UA      Ukraine     Kyiv  pressure    167   
1048498   1/1/2020    Europe      UA      Ukraine     Kyiv      pm25    158   
1048538   1/2/2020    Europe      UA      Ukraine     Kyiv      pm25    168   
1048564   1/4/2020    Europe      UA      Ukraine     Kyiv      pm25    167   

           min     max  median    variance  
4          4.5     8.0     7.0       12.84  
12         7.0    14.0    10.0       51.55  
40         5.5     9.0     7.5        9.84  
49         2.5     7.0     4.0       18.14  
62         4.5    14.5    11.5      132.58  
...        ...     ...     ...         ...  
1048252    9.8  1349.0  1010.3  1210520.00  
1048415  409.0  1119.5  1009.5   226031.00  
1048498    3.0    62.0    11.0     1717.93  
1048538   13.0   124.0    44.0     4540.99  
1048564   45.0   151.0    70.0     3673.52  

[64530 rows x 11 columns]

We note that there are 64 530 duplicated rows. In fact, there are data of some months that appear in these csv files.

Thus, we will drop the duplicated rows.


In [681]:
df.drop_duplicates(inplace = True)

In [682]:
df.shape

(3158938, 11)

As we can see, the number of rows has decreased since we dropped duplicated rows.
The number of remaining rows is correct.

#### Move the categories in the column 'Specie' to the level 1 of indexes of columns


In [683]:
# We note that the column 'Specie' contains all categories whose values per location and per day are stored in the columns: count, max, min, median, variance
df['Specie'].unique()

array(['pm10', 'humidity', 'o3', 'so2', 'wind-speed', 'wind-gust', 'dew',
       'pm25', 'no2', 'pressure', 'temperature', 'precipitation', 'co',
       'wd', 'uvi', 'aqi', 'pm1', 'mepaqi', 'pol', 'wind speed',
       'wind gust', 'neph'], dtype=object)

In [684]:
# We move the categories in the column 'Specie' to the level 1 of column index
# Thus, the level 0 of column indexes will be: count, max, min, median, variance

df = df.set_index(['Date','Continent','Country_name','Country','City','Specie'])[['count','min','max','median','variance']].unstack(5)

# After the query above, We also reset the index and convert the column 'Date', 'Country','City','Continent','Country_name' from index type to the normal column
df = df.reset_index(level = [0,1,2,3,4])

In [685]:
df.head()

Date Continent  Country_name Country          City count        \
Specie                                                           aqi    co   
0       1/1/2019    Africa      Ethiopia      ET   Addis Ababa   NaN   NaN   
1       1/1/2019    Africa       Réunion      RE   Saint-Denis   NaN   NaN   
2       1/1/2019    Africa  South Africa      ZA  Johannesburg   NaN  24.0   
3       1/1/2019    Africa  South Africa      ZA    Klerksdorp   NaN   NaN   
4       1/1/2019    Africa  South Africa      ZA    Middelburg   NaN  47.0   

                              ...      variance                     \
Specie   dew humidity mepaqi  ... precipitation    pressure    so2   
0        NaN     21.0    NaN  ...           NaN       22.92    NaN   
1       69.0    120.0    NaN  ...     1975120.0        5.39   2.03   
2        NaN     96.0    NaN  ...           0.0  1968820.00  20.92   
3        7.0     24.0    NaN  ...           0.0        0.00    NaN   
4       16.0     41.0    NaN  ...           0.0        0.00  11.47   

                                                                       
Specie temperature uvi   wd wind gust wind speed wind-gust wind-speed  
0           561.27 NaN  NaN       NaN        NaN       NaN        NaN  
1           181.08 NaN  NaN       NaN        NaN    168.15      85.86  
2           498.95 NaN  0.0       NaN        NaN     30.22       0.00  
3             0.00 NaN  0.0       NaN        NaN    211.47       0.00  
4             0.00 NaN  0.0       NaN        NaN     43.81       0.00  

[5 rows x 115 columns]

As we can see, the categories in the column 'Specie' are moved to the level 1 of index of columns.

In [686]:
df.shape

(316742, 115)

After restructuring the dataframe, there are 316 742 rows and 113 columns.

#### Convert the values in the column 'Date' into the type 'Date time'


In [687]:
df['Date'] = pd.to_datetime(df['Date'])

In [688]:
type(df.loc[1,'Date'].values[0])

pandas._libs.tslibs.timestamps.Timestamp

As we can see here, the type of values in the column 'Date' is now 'timestamp'

#### Analyse and drop unnecessary columns

We will drop the whole level 'count' of columns because these indicators is not important in our analysis

In [689]:
df.drop(columns = 'count', level = 0, inplace = True)


Then, in each index level 0, we decided to drop several columns of the index level 1 below.

In fact, we only focus on the global and individual AQI and several principal nature indicators such as temperature, humidity, wind speed, pressure.

In [690]:
col_lv1_drop = ['precipitation','mepaqi','neph','pm1','pol','dew','uvi','wd','wind gust','wind-gust']

In [691]:
df.drop(col_lv1_drop, axis = 1 , level = 1, inplace = True)

In [692]:
df.head()

Date Continent  Country_name Country          City min       \
Specie                                                          aqi   co   
0      2019-01-01    Africa      Ethiopia      ET   Addis Ababa NaN  NaN   
1      2019-01-01    Africa       Réunion      RE   Saint-Denis NaN  NaN   
2      2019-01-01    Africa  South Africa      ZA  Johannesburg NaN  3.0   
3      2019-01-01    Africa  South Africa      ZA    Klerksdorp NaN  NaN   
4      2019-01-01    Africa  South Africa      ZA    Middelburg NaN  1.0   

                           ...  variance                                    \
Specie humidity  no2   o3  ...  humidity    no2      o3     pm10      pm25   
0          12.0  NaN  NaN  ...   5154.73    NaN     NaN      NaN   3124.06   
1          55.0  0.2  4.4  ...   1601.58  20.21  122.72  1076.22   7328.02   
2           0.0  1.4  1.7  ...  15544.00  83.04   71.20  5710.13  21508.80   
3           0.0  1.0  NaN  ...      0.00   9.40     NaN   939.84   6028.24   
4           0.0  1.0  2.9  ...      0.00  40.92  126.58   521.94   3270.62   

                                                             
Specie    pressure    so2 temperature wind speed wind-speed  
0            22.92    NaN      561.27        NaN        NaN  
1             5.39   2.03      181.08        NaN      85.86  
2       1968820.00  20.92      498.95        NaN       0.00  
3             0.00    NaN        0.00        NaN       0.00  
4             0.00  11.47        0.00        NaN       0.00  

[5 rows x 53 columns]

As we can see, all the selected columns above are dropped. The number of remaining columns is correct (= 113-63).

#### Analyse missing values

Overview of missing values:

In [693]:
df.isna().sum()/df.shape[0]

              Specie     
Date                         0.000000
Continent                    0.000000
Country_name                 0.001980
Country                      0.000000
City                         0.000000
min           aqi            0.971999
              co             0.369443
              humidity       0.031748
              no2            0.188459
              o3             0.227617
              pm10           0.204087
              pm25           0.173864
              pressure       0.036948
              so2            0.307080
              temperature    0.031436
              wind speed     0.895511
              wind-speed     0.142384
max           aqi            0.971999
              co             0.369443
              humidity       0.031748
              no2            0.188459
              o3             0.227617
              pm10           0.204087
              pm25           0.173864
              pressure       0.036948
              so2       

In [694]:
df.isna().sum()

              Specie     
Date                              0
Continent                         0
Country_name                    627
Country                           0
City                              0
min           aqi            307873
              co             117018
              humidity        10056
              no2             59693
              o3              72096
              pm10            64643
              pm25            55070
              pressure        11703
              so2             97265
              temperature      9957
              wind speed     283646
              wind-speed      45099
max           aqi            307873
              co             117018
              humidity        10056
              no2             59693
              o3              72096
              pm10            64643
              pm25            55070
              pressure        11703
              so2             97265
              temperature      9957
  

As we can see in the series above, the 'aqi' columns in every index level 0 have the most important % of missing values (97%).

According to the website aqicn.org where we obtained the dataset, we can fill the missing values of AQI by taking the maximum value among those following: PM2.5, PM10, O3 (Ozone), CO (Carbon Monoxide), SO2 (Sulfur Dioxide), NO2 (Nitrogen Dioxide) 

In [695]:
#We save a copy of the original dataframe
df_copy = df.copy()

In [696]:
col_lv = ['min','max','median','variance'] #list of column labels at level 0 

In [697]:
#We use For loops that allow to replace NaN values and 0 in the columns AQI in differents groups of index. 
for n in col_lv:
    df.loc[(df[n,'aqi'].isna())|(df[n,'aqi'] ==0),[(n,'aqi')]] = df[(df[n,'aqi'].isna())|(df[n,'aqi'] ==0)][(n)][['aqi','no2','so2','pm10','pm25','o3','co']].max(axis=1)

D:\Anaconda\lib\site-packages\IPython\core\async_helpers.py:68: PerformanceWarning: indexing past lexsort depth may impact performance.
  coro.send(None)


In [698]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316742 entries, 0 to 316741
Data columns (total 53 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   (Date, )                 316742 non-null  datetime64[ns]
 1   (Continent, )            316742 non-null  object        
 2   (Country_name, )         316115 non-null  object        
 3   (Country, )              316742 non-null  object        
 4   (City, )                 316742 non-null  object        
 5   (min, aqi)               312793 non-null  float64       
 6   (min, co)                199724 non-null  float64       
 7   (min, humidity)          306686 non-null  float64       
 8   (min, no2)               257049 non-null  float64       
 9   (min, o3)                244646 non-null  float64       
 10  (min, pm10)              252099 non-null  float64       
 11  (min, pm25)              261672 non-null  float64       
 12  (min, pressure) 

The numbers of non-null values in columns AQI have increased to 312 793 rows now.

As for columns 'wind-speed' and 'wind speed', we note that they contain same type information and complete each other.

Thus, we will group all information in one column and drop the other one.

In [714]:
# Put all information about the wind speed in the column 'wind-speed' and drop the column 'wind speed'
# For loops that allow to replace NaN values in the columns wind-speed by the ones in the columns 'wind speed' 

for n in col_lv:
    df.loc[df[n,'wind-speed'].isna(),[(n,'wind-speed')]] = df[df[n,'wind-speed'].isna()][(n,'wind speed')]
    
df.drop(columns = 'wind speed', level=1, inplace = True)

D:\Anaconda\lib\site-packages\IPython\core\async_helpers.py:68: PerformanceWarning: indexing past lexsort depth may impact performance.
  coro.send(None)


In [715]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316742 entries, 0 to 316741
Data columns (total 49 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   (Date, )                 316742 non-null  datetime64[ns]
 1   (Continent, )            316742 non-null  object        
 2   (Country_name, )         316115 non-null  object        
 3   (Country, )              316742 non-null  object        
 4   (City, )                 316742 non-null  object        
 5   (min, aqi)               312793 non-null  float64       
 6   (min, co)                199724 non-null  float64       
 7   (min, humidity)          306686 non-null  float64       
 8   (min, no2)               257049 non-null  float64       
 9   (min, o3)                244646 non-null  float64       
 10  (min, pm10)              252099 non-null  float64       
 11  (min, pm25)              261672 non-null  float64       
 12  (min, pressure) 

As indicated in the detailed information above, the number of non null values in the column 'wind-speed' has increased from 271 643 to 299 435 rows.

Also, the column 'wind speed' was dropped as expected.

In [716]:
df.head()

Date Continent  Country_name Country          City   min       \
Specie                                                            aqi   co   
0      2019-01-01    Africa      Ethiopia      ET   Addis Ababa  13.0  NaN   
1      2019-01-01    Africa       Réunion      RE   Saint-Denis   8.0  NaN   
2      2019-01-01    Africa  South Africa      ZA  Johannesburg  17.0  3.0   
3      2019-01-01    Africa  South Africa      ZA    Klerksdorp   9.0  NaN   
4      2019-01-01    Africa  South Africa      ZA    Middelburg  17.0  1.0   

                           ... variance                                    \
Specie humidity  no2   o3  ...       co  humidity    no2      o3     pm10   
0          12.0  NaN  NaN  ...      NaN   5154.73    NaN     NaN      NaN   
1          55.0  0.2  4.4  ...      NaN   1601.58  20.21  122.72  1076.22   
2           0.0  1.4  1.7  ...     5.11  15544.00  83.04   71.20  5710.13   
3           0.0  1.0  NaN  ...      NaN      0.00   9.40     NaN   939.84   
4           0.0  1.0  2.9  ...    64.77      0.00  40.92  126.58   521.94   

                                                            
Specie      pm25    pressure    so2 temperature wind-speed  
0        3124.06       22.92    NaN      561.27        NaN  
1        7328.02        5.39   2.03      181.08      85.86  
2       21508.80  1968820.00  20.92      498.95       0.00  
3        6028.24        0.00    NaN        0.00       0.00  
4        3270.62        0.00  11.47        0.00       0.00  

[5 rows x 49 columns]

In [717]:
df.shape

(316742, 49)

Overall, we note that there are many rows that contain at least one missing value. These missing values will not allow us to do a efficient and accurate analyse. 

Thus, we will remove these rows and assign the new table to a dataframe called 'df_full' 

In [718]:
#filter only rows that do not contain missing values
df_full = df[~df.isna().any(axis=1)].reset_index()

In [720]:
df_full.drop(columns ='index', inplace=True)

D:\Anaconda\lib\site-packages\pandas\core\generic.py:3936: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.
  obj = obj._drop_axis(labels, axis, level=level, errors=errors)


In [721]:
df_full.shape

(120415, 49)

In [722]:
df_full.head()

Date Continent  Country_name Country            City   min       \
Specie                                                              aqi   co   
0      2019-01-01    Africa  South Africa      ZA    Johannesburg  17.0  3.0   
1      2019-01-01    Africa  South Africa      ZA      Middelburg  17.0  1.0   
2      2019-01-01    Africa  South Africa      ZA  Port Elizabeth  34.0  0.6   
3      2019-01-01    Africa  South Africa      ZA     Vereeniging   5.0  0.3   
4      2019-01-01      Asia         China      CN         Beijing  42.0  1.9   

                           ... variance                                      \
Specie humidity  no2   o3  ...       co  humidity      no2      o3     pm10   
0           0.0  1.4  1.7  ...     5.11  15544.00    83.04   71.20  5710.13   
1           0.0  1.0  2.9  ...    64.77      0.00    40.92  126.58   521.94   
2           0.0  0.5  0.9  ...     0.00  13379.40    29.34    0.00  3152.75   
3           0.0  0.5  2.9  ...    71.32  14541.90    43.96   41.73  3960.36   
4          18.0  3.2  0.5  ...   144.93   2362.28  1207.58  570.75  5678.14   

                                                             
Specie      pm25    pressure     so2 temperature wind-speed  
0       21508.80  1968820.00   20.92      498.95       0.00  
1        3270.62        0.00   11.47        0.00       0.00  
2        6353.78        0.00   77.77     1360.49       0.00  
3       17139.20  1677650.00   18.10      400.22      10.10  
4       17289.00       15.28  254.59      269.91      41.68  

[5 rows x 49 columns]

Based on the new dataframe 'df_full', we note that the number of locations is too big that we cannot take into account 100% of locations in our analysis. In addition, the data reported is not completed for a big part of locations (number of days reported < 627 days).


These are reasons why we will reduce the location sample for our analysis by following these steps below:

    * Examine the data contribution by each location by observing the number of rows of each location (country, city)

    * Remove cities that have less than 550 rows. In fact, 550 rows correspond to the number of days from 1.1.2019 to 3.7.2020
    
    * Aggregate cities by country and continent, then select several locations for further analysis.

In [723]:
# Groupby location (country, city) in order to count the rows and we assign this table to a variable

count_row = df_full.groupby(['Continent','Country_name','Country','City']).size().sort_values(ascending=False).reset_index(name='Count')

In [724]:
# Extract the list of cities whose values were reported for more than 550 days

count_row[count_row['Count'] > 550]

Continent    Country_name Country         City  Count
0             Asia        Thailand      TH   Chiang Mai    618
1    South America          Brazil      BR   São Paulo    614
2             Asia        Thailand      TH      Bangkok    613
3             Asia           India      IN        Delhi    612
4    North America          Mexico      MX  Mexico City    612
..             ...             ...     ...          ...    ...
97          Europe  United Kingdom      GB      Cardiff    557
98          Europe          Poland      PL      Kraków    554
99   North America   United States      US    Las Vegas    553
100  North America   United States      US       Tucson    552
101           Asia           Japan      JP      Niigata    551

[102 rows x 5 columns]

By examining the list above, we note that there are 101 cities whose date was reported for at least 550 days. 

Given that this number of locations is still big, we decide to only take 20 cities from 17 countries of 4 continents that contain a sufficient amount of data for our analysis:

In [764]:
df_test=df_full.replace({'City':'WrocÅ\x82aw'},{'City':'Wrocław'})

In [768]:
# df_test[df_test['City']=='Wrocław']

df_test['City'].unique()

array(['Johannesburg', 'Middelburg', 'Port Elizabeth', 'Vereeniging',
       'Beijing', 'Changchun', 'Changsha', 'Chengdu', 'Chongqing',
       'Fushun', 'Fuzhou', 'Guangzhou', 'Guiyang', 'Haikou', 'Harbin',
       'Hefei', 'Hegang', 'Jieyang', 'Jinan', 'Kunming', 'Lanzhou',
       'Lhasa', 'Nanchang', 'Nanjing', 'Nanning', 'Qingdao',
       'Qinhuangdao', 'Qiqihar', 'Shanghai', 'Shantou', 'Shenyang',
       'Shenzhen', 'Shijiazhuang', 'Taiyuan', 'Wuhan', 'Xiamen', 'Xining',
       'Xinxiang', 'Xiâ\x80\x99an', 'Xuchang', 'Yinchuan', 'Zhengzhou',
       'Zhuzhou', 'Ã\x9crümqi', 'Nicosia', 'Hong Kong', 'Bengaluru',
       'Bhopal', 'Delhi', 'GhÄ\x81ziÄ\x81bÄ\x81d', 'Hyderabad',
       'HÄ\x81pur', 'Jaipur', 'Kolkata', 'Mumbai', 'Muzaffarnagar',
       'Thiruvananthapuram', 'Visakhapatnam', 'Karaj', 'Sanandaj',
       'Tabriz', 'Tehran', 'Haifa', 'Petaáº\x96 Tiqwa', 'Tel Aviv',
       'Akita', 'Chiba', 'Fukuoka', 'Gifu-shi', 'Hiroshima', 'Kagoshima',
       'Kanazawa', 'Kobe', 'Kochi', '

In [725]:
dct_sample = {'Mexico City':'MX','Guadalajara':'MX','Vancouver':'CA','Las Vegas':'US','Jacksonville':'US','São Paulo':'BR','Concepción':'CL','Bogotá':'CO','Sydney':'AU','Brisbane':'AU','Bangkok':'TH','Delhi':'IN','Taipei':'TW','Seoul':'KR','Budapest':'HU','Wrocław':'PL','London':'GB','Hamburg':'DE','Brussels':'BE','Rijeka':'HR'}

In [733]:
df_full['City'].unique()

array(['Johannesburg', 'Middelburg', 'Port Elizabeth', 'Vereeniging',
       'Beijing', 'Changchun', 'Changsha', 'Chengdu', 'Chongqing',
       'Fushun', 'Fuzhou', 'Guangzhou', 'Guiyang', 'Haikou', 'Harbin',
       'Hefei', 'Hegang', 'Jieyang', 'Jinan', 'Kunming', 'Lanzhou',
       'Lhasa', 'Nanchang', 'Nanjing', 'Nanning', 'Qingdao',
       'Qinhuangdao', 'Qiqihar', 'Shanghai', 'Shantou', 'Shenyang',
       'Shenzhen', 'Shijiazhuang', 'Taiyuan', 'Wuhan', 'Xiamen', 'Xining',
       'Xinxiang', 'Xiâ\x80\x99an', 'Xuchang', 'Yinchuan', 'Zhengzhou',
       'Zhuzhou', 'Ã\x9crümqi', 'Nicosia', 'Hong Kong', 'Bengaluru',
       'Bhopal', 'Delhi', 'GhÄ\x81ziÄ\x81bÄ\x81d', 'Hyderabad',
       'HÄ\x81pur', 'Jaipur', 'Kolkata', 'Mumbai', 'Muzaffarnagar',
       'Thiruvananthapuram', 'Visakhapatnam', 'Karaj', 'Sanandaj',
       'Tabriz', 'Tehran', 'Haifa', 'Petaáº\x96 Tiqwa', 'Tel Aviv',
       'Akita', 'Chiba', 'Fukuoka', 'Gifu-shi', 'Hiroshima', 'Kagoshima',
       'Kanazawa', 'Kobe', 'Kochi', '

In [726]:
# We create a function to copy all data relating to the selected locations and paste them into a new dataframe called "df_sample"
df_sample = []

for city,code in dct_sample.items():
    data = df_full[(df_full['Country'] == code) & (df_full['City'] == city)]
    df_sample.append(data)

df_sample = pd.concat(df_sample)


In [727]:
df_sample.shape

(9492, 49)

The number of rows is correct. 

In [729]:
df_sample.groupby('City').size().sort_values(ascending = False)

City
Bangkok         613
Mexico City     612
Delhi           612
Taipei          609
Budapest        607
London          605
Seoul           603
Hamburg         599
Guadalajara     595
Vancouver       594
Jacksonville    594
Sydney          590
Rijeka          578
Brisbane        578
Las Vegas       553
Brussels        550
dtype: int64

### Exploratory Data Analysis

Remind: We use our dataframe called "df_sample" in this part

In [669]:
df_sample.head()

Date      Continent Country_name Country         City   min       \
Specie                                                               aqi   co   
192    2019-01-01  North America       Mexico      MX  Mexico City  17.0  1.2   
424    2019-01-10  North America       Mexico      MX  Mexico City  17.0  1.2   
588    2020-01-10  North America       Mexico      MX  Mexico City   2.0  1.2   
808    2019-01-11  North America       Mexico      MX  Mexico City   5.0  1.2   
973    2020-01-11  North America       Mexico      MX  Mexico City   5.0  1.2   

                           ... variance                                       \
Specie humidity  no2   o3  ...       co humidity      no2       o3      pm10   
192         0.0  0.2  0.8  ...   576.52  3598.95  1121.85  9284.55  28058.40   
424         0.0  0.1  0.8  ...   197.30  4286.00  1188.42  3645.11   2818.12   
588        17.0  1.9  0.1  ...   586.29  3639.64  2261.62  2413.64   5760.66   
808         0.0  0.8  0.1  ...   305.46  4499.28  1235.21  6601.94   3752.21   
973        17.0  1.9  0.1  ...   264.19  2224.42  1922.62  5841.12   4602.23   

                                                             
Specie      pm25   pressure      so2 temperature wind-speed  
192     35627.10   63439.20   964.49      311.83      36.38  
424      6069.75   72415.70  1150.62      197.35      25.54  
588     10694.60    9299.27   592.31      291.96      86.81  
808      4874.76  138807.00   658.22      168.77      10.96  
973      9478.18    7006.84   651.72      226.29      60.68  

[5 rows x 49 columns]